### PR TITLE
Reduce direct memory overhead per EpollEventLoop when using EpollData…

### PR DESCRIPTION
**Pull request by normanmaurer**
…gramChannel

Motivation:

When using a linux distribution that supports sendmmsg(...) we allocated enough direct memory per EpollEventLoop to be able to write IOV_MAX number of iovecs per message that can be written per sendmmsg.
The number of messages that can be written per sendmmsg(...) call is limited by UIO_MAX_IOV.

In practice this resulted in an allocation of 16MB direct memory per EpollEventLoop instance that stayed allocated until the EpollEventLoop was shutdown which happens as part of the shutdown of the enclosing EpollEVentLoopGroup.

This resulted in quite some heavy direct memory usage in practice even when in practice we have very slim changes to ever need all of the memory.

Modification:

Adjust NativeDatagramPacketArray to share one IovArray instance across all NativeDatagramPacket instances it holds. This limits the max number of iovecs we can write across all messages to IOV_MAX per sendmmsg(...) call.
This in practice will still be enough to allow us to write multiple messages with one syscall while keep the memory overhead to a minimum.

Result:

Smaller direct memory footprint per EpollEventLoop when using EpollDatagramChannel on distributions that support sendmmsg(...).
Fixes https://github.com/netty/netty/issues/8814

### CODE REVIEW

**Comment by normanmaurer** [comment]
/cc @njhill 

@johnou this should fix your problem

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/NativeDatagramPacketArray.java [hunk 1]
```diff
@@ -85,24 +95,18 @@ int count() {
 
     void clear() {
         this.count = 0;
+        this.iovArray.clear();
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/NativeDatagramPacketArray.java]: no... We deallocate the direct memory of the IovArray once it is destroyed only. The IovArray itself has something like 16kb memory need, so nothing to worry about.

[the remark this answers is not in the document]
Comment (johnou) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/NativeDatagramPacketArray.java]: Oops right not in clean that just resets the count etc. when I looked I didn't see any destroy method.

[the remark this answers is not in the document]
Comment (normanmaurer) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/NativeDatagramPacketArray.java]: no worries :)

**Comment by normanmaurer** [comment]
@johnou did you have a chance yet to test this ?

**Comment by johnou** [comment]
@normanmaurer yes, and things are looking good. Just wanted to run one of our load tests first.

**Comment by normanmaurer** [comment]
@johnou <3... keep me posted

**Comment by johnou** [comment]
@normanmaurer I can already say that the initial memory footprint is much better now :ok_hand:

**Comment by normanmaurer** [comment]
@johnou ok let me know if there is any change after running it for a bit.

**Comment by johnou** [comment]
@normanmaurer some anomalies with the load tests which I need to investigate, going to need a bit more time to dig into what's up.

**Comment by johnou** [comment]
@normanmaurer yeah it's something else which I just reproduced with nio and Netty 4.1.30, so green light from me, thanks a lot!

**Comment by johnou** [comment]
@normanmaurer not sure if you want a new issue opened or it's expected behaviour but when I was testing with NIO the app failed to bootstrap :)

```
Caused by: java.lang.IllegalStateException: failed to create a new resolver
	at io.netty.resolver.AddressResolverGroup.getResolver(AddressResolverGroup.java:71) ~[netty-resolver-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap.doResolveAndConnect0(Bootstrap.java:200) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap.access$000(Bootstrap.java:49) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:188) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:174) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:511) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:485) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:424) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:103) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetSuccess(AbstractChannel.java:982) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:516) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:427) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:486) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:495) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at com.sulake.h4k.common.util.io.NonBlockingIOThreadFactory.lambda$newThread$0(NonBlockingIOThreadFactory.java:18) ~[h4k-common-develop-SNAPSHOT.jar:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at com.sulake.h4k.common.util.io.NonBlockingIOThreadFactory.lambda$newThread$1(NonBlockingIOThreadFactory.java:26) ~[h4k-common-develop-SNAPSHOT.jar:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_191]
Caused by: io.netty.channel.ChannelException: Unable to create Channel from class class io.netty.channel.socket.nio.NioDatagramChannel
	at io.netty.channel.ReflectiveChannelFactory.newChannel(ReflectiveChannelFactory.java:46) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.AbstractBootstrap.initAndRegister(AbstractBootstrap.java:320) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.AbstractBootstrap.register(AbstractBootstrap.java:234) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsNameResolver.<init>(DnsNameResolver.java:395) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsNameResolverBuilder.build(DnsNameResolverBuilder.java:430) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsAddressResolverGroup.newNameResolver(DnsAddressResolverGroup.java:114) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsAddressResolverGroup.newResolver(DnsAddressResolverGroup.java:94) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsAddressResolverGroup.newResolver(DnsAddressResolverGroup.java:79) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.AddressResolverGroup.getResolver(AddressResolverGroup.java:69) ~[netty-resolver-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap.doResolveAndConnect0(Bootstrap.java:200) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap.access$000(Bootstrap.java:49) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:188) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:174) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:511) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:485) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:424) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:103) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetSuccess(AbstractChannel.java:982) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:516) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:427) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:486) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:495) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at com.sulake.h4k.common.util.io.NonBlockingIOThreadFactory.lambda$newThread$0(NonBlockingIOThreadFactory.java:18) ~[h4k-common-develop-SNAPSHOT.jar:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at com.sulake.h4k.common.util.io.NonBlockingIOThreadFactory.lambda$newThread$1(NonBlockingIOThreadFactory.java:26) ~[h4k-common-develop-SNAPSHOT.jar:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_191]
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.GeneratedConstructorAccessor31.newInstance(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_191]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_191]
	at io.netty.channel.ReflectiveChannelFactory.newChannel(ReflectiveChannelFactory.java:44) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.AbstractBootstrap.initAndRegister(AbstractBootstrap.java:320) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.AbstractBootstrap.register(AbstractBootstrap.java:234) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsNameResolver.<init>(DnsNameResolver.java:395) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsNameResolverBuilder.build(DnsNameResolverBuilder.java:430) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsAddressResolverGroup.newNameResolver(DnsAddressResolverGroup.java:114) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsAddressResolverGroup.newResolver(DnsAddressResolverGroup.java:94) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsAddressResolverGroup.newResolver(DnsAddressResolverGroup.java:79) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.AddressResolverGroup.getResolver(AddressResolverGroup.java:69) ~[netty-resolver-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap.doResolveAndConnect0(Bootstrap.java:200) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap.access$000(Bootstrap.java:49) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:188) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:174) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:511) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:485) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:424) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:103) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetSuccess(AbstractChannel.java:982) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:516) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:427) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:486) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:495) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at com.sulake.h4k.common.util.io.NonBlockingIOThreadFactory.lambda$newThread$0(NonBlockingIOThreadFactory.java:18) ~[h4k-common-develop-SNAPSHOT.jar:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at com.sulake.h4k.common.util.io.NonBlockingIOThreadFactory.lambda$newThread$1(NonBlockingIOThreadFactory.java:26) ~[h4k-common-develop-SNAPSHOT.jar:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_191]
Caused by: io.netty.channel.ChannelException: Failed to open a socket.
	at io.netty.channel.socket.nio.NioDatagramChannel.newSocket(NioDatagramChannel.java:88) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.socket.nio.NioDatagramChannel.<init>(NioDatagramChannel.java:116) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at sun.reflect.GeneratedConstructorAccessor31.newInstance(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_191]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_191]
	at io.netty.channel.ReflectiveChannelFactory.newChannel(ReflectiveChannelFactory.java:44) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.AbstractBootstrap.initAndRegister(AbstractBootstrap.java:320) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.AbstractBootstrap.register(AbstractBootstrap.java:234) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsNameResolver.<init>(DnsNameResolver.java:395) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsNameResolverBuilder.build(DnsNameResolverBuilder.java:430) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsAddressResolverGroup.newNameResolver(DnsAddressResolverGroup.java:114) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsAddressResolverGroup.newResolver(DnsAddressResolverGroup.java:94) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsAddressResolverGroup.newResolver(DnsAddressResolverGroup.java:79) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.AddressResolverGroup.getResolver(AddressResolverGroup.java:69) ~[netty-resolver-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap.doResolveAndConnect0(Bootstrap.java:200) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap.access$000(Bootstrap.java:49) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:188) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:174) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:511) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:485) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:424) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:103) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetSuccess(AbstractChannel.java:982) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:516) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:427) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:486) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:495) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at com.sulake.h4k.common.util.io.NonBlockingIOThreadFactory.lambda$newThread$0(NonBlockingIOThreadFactory.java:18) ~[h4k-common-develop-SNAPSHOT.jar:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at com.sulake.h4k.common.util.io.NonBlockingIOThreadFactory.lambda$newThread$1(NonBlockingIOThreadFactory.java:26) ~[h4k-common-develop-SNAPSHOT.jar:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_191]
Caused by: java.net.SocketException: maximum number of DatagramSockets reached
	at sun.net.ResourceManager.beforeUdpCreate(ResourceManager.java:73) ~[?:1.8.0_191]
	at sun.nio.ch.DatagramChannelImpl.<init>(DatagramChannelImpl.java:111) ~[?:1.8.0_191]
	at sun.nio.ch.SelectorProviderImpl.openDatagramChannel(SelectorProviderImpl.java:42) ~[?:1.8.0_191]
	at io.netty.channel.socket.nio.NioDatagramChannel.newSocket(NioDatagramChannel.java:86) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.socket.nio.NioDatagramChannel.<init>(NioDatagramChannel.java:116) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at sun.reflect.GeneratedConstructorAccessor31.newInstance(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_191]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[?:1.8.0_191]
	at io.netty.channel.ReflectiveChannelFactory.newChannel(ReflectiveChannelFactory.java:44) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.AbstractBootstrap.initAndRegister(AbstractBootstrap.java:320) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.AbstractBootstrap.register(AbstractBootstrap.java:234) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsNameResolver.<init>(DnsNameResolver.java:395) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsNameResolverBuilder.build(DnsNameResolverBuilder.java:430) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsAddressResolverGroup.newNameResolver(DnsAddressResolverGroup.java:114) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsAddressResolverGroup.newResolver(DnsAddressResolverGroup.java:94) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.dns.DnsAddressResolverGroup.newResolver(DnsAddressResolverGroup.java:79) ~[netty-resolver-dns-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.resolver.AddressResolverGroup.getResolver(AddressResolverGroup.java:69) ~[netty-resolver-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap.doResolveAndConnect0(Bootstrap.java:200) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap.access$000(Bootstrap.java:49) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:188) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:174) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:511) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:485) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:424) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:103) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetSuccess(AbstractChannel.java:982) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:516) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:427) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:486) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:495) ~[netty-transport-4.1.33.Final.jar:4.1.33.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at com.sulake.h4k.common.util.io.NonBlockingIOThreadFactory.lambda$newThread$0(NonBlockingIOThreadFactory.java:18) ~[h4k-common-develop-SNAPSHOT.jar:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at com.sulake.h4k.common.util.io.NonBlockingIOThreadFactory.lambda$newThread$1(NonBlockingIOThreadFactory.java:26) ~[h4k-common-develop-SNAPSHOT.jar:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.33.Final.jar:4.1.33.Final]
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_191]
```

**Comment by johnou** [comment]
I guess epoll survives because it uses `sendmmsg(...)`..

**Comment by normanmaurer** [comment]
@johnou how many EventLoops you are using here ? Also are you sure you call `AddressResolverGroup.close()` once you are done using it ? It looks like you may miss to do this and so not close datagram sockets.

**Comment by johnou** [comment]
@normanmaurer that's part of Redisson, I will dig through the code later, 16 I believe.

**Comment by normanmaurer** [comment]
@johnou ok let me know if you find out anything or if you need any more help

**Comment by johnou** [comment]
@normanmaurer the reason NIO was failing is because we are running a security manager.. which limits udp sockets to 25 :man_facepalming: we worked around it by setting sun.net.maxDatagramSockets to 1024.

**Comment by normanmaurer** [comment]
Makes sense. Thanks for closing the loop

**Comment by johnou** [comment]
Just to be clear it's the JVM which imposes this limit if a security manager is installed, not something we intentionally configured.

**Pull request merged**